Add epoch stuff to the hyparview peer service manager

junghunyoo opened the pull request.

junghunyoo commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -462,7 +478,9 @@ handle_info({'EXIT', From, _Reason},
 
 handle_info({connected, Peer, Tag, _RemoteState}, State) ->
-    handle_connect(Peer, Tag, State);
+    %% @todo retreive the PeerEpoch from the Peer.
```
The information which a peer can provide is a set of the active view. In order to handle the join properly, we need to find a method to pass the epoch value to the `partisan_peer_service_server` (similar to the `get_local_state()` in the `partisan_peer_service_manager`).

cmeiklejohn replied:
Can't you just expand the tuple like we did when we added ```Tag```?

(remote state is going away as soon as I have time to remove it.)

Can you change the message to `{connected, Peer, Tag, Epoch, RemoteState}`?

junghunyoo replied:
In order the change the message to contain the `Epoch`, the `get_local_state()` needs to return the epoch value as well. Currently it only returns a set of active view. Can I change the `get_local_state()` function? It looks like that function is an interface, so it can affect other managers as well.

junghunyoo commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -531,23 +569,29 @@ handle_message({neighbor_accepted, Peer, Tag, Sender},
+                    %% Add to active view.
+                    add_to_active_view(Peer, Tag, State1);
+                false ->
+                    %% @todo Need to send the disconnect message here?
```
Before a node sent the `neighbor_accepted` message to the peer, it added the peer to its active view. So, if the message is not valid for adding, I think we need to let the node know this and it can remove the peer from the active view.

cmeiklejohn replied:
Can we just send a disconnect here?

junghunyoo replied:
I will check whether sending the disconnect message here is fine.

junghunyoo commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -556,30 +600,45 @@ handle_message({neighbor_accepted, Peer, Tag, Sender},
     {reply, ok, State};
 
 handle_message({neighbor_rejected, Peer, _Sender}, State) ->
+    %% @todo Is this necessary?
```
As I know, we did not add the peer to the active view before sending the `neighbor_request` message. I am not sure whether this trigger is needed.

cmeiklejohn replied:
I think this section is mainly here for logging that it attempted to peer and was rejected, but, we do need to disconnect the node because we had to open a connection to it to send the request, right?

junghunyoo replied:
Yes. I did not consider the opened connection. I will remove this comment.

cmeiklejohn replied:
I think you need to add code to close the open connection, right?

junghunyoo replied:
As I know, the `gen_server:cast(?MODULE, {disconnect, Peer})` will close the open connection.

junghunyoo commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -491,37 +509,57 @@ code_change(_OldVsn, State, _Extra) ->
+                    %% Add node into the active view.
+                    add_to_active_view(Peer, Tag, State0);
+                false ->
+                    %% @todo Need to send the disconnect message here?
```
Similar to the `neighbor_accepted` message, a node added the peer to its active view before sending the `neighbor` message. So, I think we need to let the node know that the `neighbor` message fails to add the node to the peer's active view.

junghunyoo commented:
After applying this change, I saw only one symmetric test failure (more than 30 times test run) and it was the correct result (it ran the symmetric checking before a node receives the `disconnect` message).
While I was testing, I rarely saw the `hyparview_manager_low_active_test` failed with `Graph is not connected!` error message (similar to the travis-ci/pr failure in this PR). I think this is not a real failure since some nodes can be partitioned, but I am not sure about this.

cmeiklejohn commented:
Can you merge this with the latest master?

cmeiklejohn commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -57,14 +57,19 @@
 -type suspected() :: sets:set(node_spec()).
 -type reserved() :: dict:dict(atom(), node_spec()).
 -type tag() :: atom().
+-type epoch() :: non_neg_integer().
```
Can you put a comment that says why `epoch` and `epoch_count` are different?

junghunyoo replied:
I will add the comment.

cmeiklejohn commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -57,14 +57,19 @@
 %% temporary exceptions
--export([delete_state_from_disk/0]).
+-export([delete_state_from_disk/0,
+         delete_epoch_from_disk/0]).
```
Why is epoch not part of the state?

junghunyoo replied:
I assume that it is specifically for the sets of active & passive views, but if it is okay to handle them together I will change it.

cmeiklejohn replied:
State can be anything, and it might make debugging easier if you synchronize them together: just an idea.

cmeiklejohn replied:
(it also cuts down on repeated code for writing/reading from disk, too)

cmeiklejohn commented:
Can you modify the tests as well to work with a larger set of peers to verify it still works then?  The divergence branch of Lasp had trouble getting a connected graph with 8 - 16 nodes.

junghunyoo commented:
I will merge the master branch, once we are happy with new changes.

cmeiklejohn commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -288,8 +306,9 @@ handle_call(members, _From, #state{active=Active}=State) ->
-    {reply, {ok, Active}, State};
+handle_call(get_local_state, _From, #state{active=Active,
+                                           epoch=Epoch}=State) ->
+    {reply, {ok, [Active, Epoch]}, State};
```
Can this be a tuple instead of a list?

cmeiklejohn commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -477,8 +496,8 @@ handle_info({'EXIT', From, _Reason},
 
-handle_info({connected, Peer, Tag, _RemoteState}, State) ->
-    handle_connect(Peer, Tag, State);
+handle_info({connected, Peer, Tag, PeerEpoch, _RemoteState}, State) ->
```
Did you make a similar change in the other peer service managers?

cmeiklejohn commented on src/partisan_peer_service_client.erl:
```diff
@@ -143,7 +143,12 @@ decode(Message) ->
 handle_message({state, Tag, LocalState},
                #state{peer=Peer, from=From}=State) ->
-    From ! {connected, Peer, Tag, LocalState},
+    case LocalState of
```
I'm not sure I like this change.

cmeiklejohn replied:
Disregard.

cmeiklejohn commented on src/partisan_hyparview_peer_service_manager.erl:
```diff
@@ -507,37 +526,57 @@ code_change(_OldVsn, State, _Extra) ->
+                    %% Add node into the active view.
+                    add_to_active_view(Peer, TheirTag, State0);
+                false ->
+                    %% @todo Need to send the disconnect message here?
```
What do we need to do here?  What makes us get into this situation?

junghunyoo replied:
Before sending the neighbor message, the functions (like `perform_join()`) add the peer to their active view. So, if the neighbor message fails to add the peer, I think we need to let the peer know that neighboring was failed.

cmeiklejohn commented on test/partisan_SUITE.erl:
```diff
@@ -38,7 +38,7 @@
 
 -define(APP, partisan).
--define(CT_SLAVES, [rita, sue, bob, jerome]).
+-define(CLIENT_NUMBER, 3).
```
Does this test still pass if you run with 8 - 16?

junghunyoo replied:
Yes, but you need to increase the waiting time to make the cluster stable. If test checks the symmetric status in the middle of exchanging join messages, it will fail.

cmeiklejohn replied:
It might be useful to not have to use a timer and have a function that we can wait on that returns a stability metric or something.

junghunyoo replied:
I agree with that.

cmeiklejohn replied:
Do you have any ideas how we could add that here to remove the sleeps?

junghunyoo replied:
I think we can run the symmetric checking codes after all nodes' forward_join messages finish the random walk, but I am not sure how we can check it yet.

cmeiklejohn commented:
Can you explain what the last few changes do?

junghunyoo commented:
Those changes are for preventing the race conditions I found:

- A peer should not send the `neighbor` message back if a node is already in the active view after receiving the `forward_join` message.

- When we drop a peer, we send the `disconnect` message to peer and trigger the disconnection asynchronously. Sometimes, we can receive the message for adding the peer to the active view before we disconnect the peer. This means that a newly re-added peer can be removed by previous disconnect trigger. So, I added a barrier to check whether the peer is okay to disconnect.

- When we drop a peer, we send the `disconnect` message and trigger the disconnection in the same function. This can cause the disconnection before sending the message. So I added the `disconnect_ack` message and a node will trigger the disconnection after receiving the `disconnect_ack` message not during the dropping.

cmeiklejohn commented:
```
When we drop a peer, we send the disconnect message to peer and trigger the disconnection asynchronously. Sometimes, we can receive the message for adding the peer to the active view before we disconnect the peer. This means that a newly re-added peer can be removed by previous disconnect trigger. So, I added a barrier to check whether the peer is okay to disconnect.
```

Isn't this what the epoch was suppose to fix?  What additional work was needed?

junghunyoo commented:
The epoch stuff assume that sending the disconnect message and disconnecting are a single atomic behaviour, but in our implementation it can be interrupted.
